rholang/examples/linking/tests/LinkedList_test.rho

```diff
@@ -0,0 +1,265 @@
+          match [b1, b2] with [true, true] => {
+            return(true)
+          }
+          _ => { return(false) }
+        } |
+        contract emptyList_test(_) = {
+          //nothing much to test here, should just be the empty list constant
+          match emptyList == emptyList with true => { "EmptyList: Pass".display("\n") }
+          _ => { "EmptyList: Fail".display("\n") }
+        } |
+        contract head_test(_) = {
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(list <- NonEmptyList) {
+              new result1, result2, answer1, answer2 in {
+                head(list, result1) | head(emptyList, result2) |
+                answer1(1) | answer2(Nil) |
+                for(r1 <- result1; r2 <- result2; a1 <- answer1; a2 <- answer2) {
+                  match [r1 == a1, r2 == a2] with [true, true] => {
+                    "Head: Pass".display("\n")
+                  }
+                  _ => { "Head: Fail".display("\n") }
+                }
+              }
+            }
+          }
+        } |
+        contract tail_test(_) = {
+          //return the tail of the list
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(tl <- NonEmptyList) {
+              prepend(2, tl, NonEmptyList) | for(list <- NonEmptyList) {
+                new result1, result2, answer1, answer2 in {
+                  tail(list, result1) | tail(emptyList, result2) |
+                  answer1(tl) | answer2(Nil) |
+                  for(r1 <- result1; r2 <- result2; a1 <- answer1; a2 <- answer2) {
+                    match [r1 == a1, r2 == a2] with [true, true] => {
+                      "Tail: Pass".display("\n")
+                    }
+                    _ => { "Tail: Fail".display("\n") }
+                  }
+                }
+              }
+            }
+          }
+        } |
+        contract prepend_test(_) = {
+          //create a new list with the supplied head attached
+          //to the supplied tail
+          new hd, tl, result in {
+            hd(1) | tl(emptyList) |
+            for(h <- hd; l <- tl) {
+              prepend(h, l, result) | for(r <- result) {
+                head(r, hd) | tail(r, tl) | 
+                for(rh <- hd; rl <- tl) {
+                  match [rh == h, rl == l] with [true, true] => {
+                    "Prepend: Pass".display("\n")
+                  }
+                  _ => { "Prepend: Fail".display("\n") }
+                }
+              }
+            }
+          }
+        } |
+        contract fold_test(_) = {
+          //left associative fold
+          new sum, rList, result in {
+            contract sum(head, accVal, return) = {
+              return(head + accVal)
+            } |
+            range(1, 10, rList) | for(list <- rList) {
+              fold(list, 0, sum, result) |
+              for(r <- result) {
+                match r == 45 with true => { "Fold: Pass".display("\n") }
```

Comment:
I'm a bit confused here

```
>>> 1+2+3+4+5+6+7+8+9+10
55
```

Is that not what this test does?

Reply:
Oh the 10 isn't included in the range.

Reply:
I guess that is consistent with what I'm used to in Python.

```
>>> sum(range(0,10))
45
```

Reply:
It only goes up to 9, the end limit is excluded